lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -2,271 +2,521 @@
-//
-// ## Step #2: Allocate storage
-//
-// In reverse order (PO), allocate the parent storage object for each opaque
-// value.
-//
-// [REUSE] If storage has already been allocated for the current live range,
-// then simply reuse it.
-//
-// If the value's use composes a parent object from this value, and use's
-// storage can be projected from, then mark the value's storage as a projection
-// from the use value. [REUSE] Also inherit the use's storage identifier, and
-// add an interval to the live range with the current projection path.
-//
-// A use can be projected from if its allocation is available at (dominates)
-// this value and using the same storage over the interval from this value to
-// the use does not overlap with the existing live range.
-//
-// Checking interference requires checking all operands that have been marked as
-// projections. In the case of block arguments, it means checking the terminator
-// operands of all predecessor blocks.
-//
-// [REUSE] Rather than checking all value operands, each live range will contain
-// a set of intervals. Each interval will be associated with a projection path.
-//
-// Opaque value's that are the root of all projection paths now have their
-// `storageAddress` assigned to an `alloc_stack` or argument. Opaque value's
-// that are projections do not yet have a `storageAddress`.
-//
-// ## Step #3. Rewrite opaque values
-//
-// In forward order (RPO), rewrite each opaque value definition, and all its
-// uses. This generally involves creating a new `_addr` variant of the
-// instruction and obtaining the storage address from the `valueStorageMap`.
-//
-// If this value's storage is a projection of the value defined by its composing
-// use, then first generate instructions to materialize the projection. This is
-// a recursive process starting with the root of the projection path.
-//
-// A projection path will be materialized once, for the leaf subobject. When
-// this happens, the `storageAddress` will be assigned for any intermediate
-// projection paths. When those values are rewritten, their `storageAddress`
-// will already be available.
-//
+//===----------------------------------------------------------------------===//
+///
+/// This pass removes "opaque SILValues" by translating them into addressable
+/// memory locations such as a stack locations. This is mandatory for IRGen.
+///
+/// Lowering to LLVM IR requires each SILValue's type to be a valid "SIL storage
+/// type". Opaque SILValues have address-only types. Address-only values require
+/// indirect storage in LLVM, so their SIL storage type must be an address type.
+///
+/// This pass should not introduce any semantic copies. Guaranteed values always
+/// reuse the borrowed value's storage. This means that we SIL cannot allow
+/// guaranteed opaque uses unless they are projections of the definition. In
+/// particular, borrowed structs, tuples, and enums of address-only types are
+/// not allowed.
```

Comment:
SIL.rst
```
Forwarding Address-Only Values
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Address-only values are potentially unmovable when borrowed. This
means that they cannot be forwarded with guaranteed ownership unless
the forwarded value has the same representation as in the original
value and can reuse the same storage. Non-destructive projection is
allowed, such as `struct_extract`. Aggregation, such as `struct`, and
destructive disaggregation, such as `switch_enum` is not allowed. This
is an invariant for OSSA with opaque SIL values for these reasons:

1. To avoid implicit semantic copies. For move-only values, this allows
complete diagnostics. And in general, it makes it impossible for SIL
passes to "accidentally" create copies.

2. To reuse borrowed storage. This allows the optimizer to share the same
storage for multiple exclusive reads of the same variable, avoiding
copies. It may also be necessary to support native Swift atomics, which
will be unmovable-when-borrowed.
```
SILVerifier::checkOwnershipForwardingInst
```
    // Address-only values are potentially unmovable when borrowed. Ensure that
    // guaranteed address-only values are forwarded with the same
    // representation. Non-destructive projection is allowed. Aggregation and
    // destructive disaggregation is not allowed. See SIL.rst, Forwarding
    // Addres-Only Values.
    if (ownership == OwnershipKind::Guaranteed
        && OwnershipForwardingMixin::isAddressOnly(i)) {
      require(OwnershipForwardingMixin::hasSameRepresentation(i),
              "Forwarding a guaranteed address-only value requires the same "
              "representation since no move or copy is allowed.");
    }
```
SILVerifier::visitSILPhiArgument
```
        // Address-only values are potentially unmovable when borrowed. See also
        // checkOwnershipForwardingInst. A phi implies a move of its arguments
        // because they can't necessarilly all reuse the same storage.
        require((!arg->getType().isAddressOnly()
                 || arg->getOwnershipKind() != OwnershipKind::Guaranteed) &&
                "Guaranteed address-only phi not allowed--implies a copy");
```
AddressLowering file comment
```
/// This pass removes "opaque SILValues" by translating them into addressable
/// memory locations such as a stack locations. This is mandatory for IRGen.
///
/// Lowering to LLVM IR requires each SILValue's type to be a valid "SIL storage
/// type". Opaque SILValues have address-only types. Address-only values require
/// indirect storage in LLVM, so their SIL storage type must be an address type.
///
/// This pass never creates copies except to replace explicit value copies
/// (copy_value, load [copy], store). For move-only values, this allows complete
/// diagnostics. And in general, it makes it impossible for SIL passes to
/// "accidentally" create copies.
///
/// This pass inserts moves (copy_addr [take] [initialize]) of owned values to
/// - compose aggregates
/// - resolve phi interference
///
/// For guarantee values, this pass inserts neither copies nor
/// moves. Address-only values are potentially unmovable when borrowed. This
/// means that guaranteed address-only aggregates and phis are prohibited. This
/// SIL invariant is enforced by SILVerifier::checkOwnershipForwardingInst() and
/// SILVerifier::visitSILPhiArgument().
///
/// ## Step #1: Map opaque values
```
@rjmccall have I explained this well enough?